src/user-event/press/press.ts

```diff
@@ -1,16 +1,118 @@
+}
 
-  await wait(this.config);
-  dispatchHostEvent(element, 'press', EventBuilder.Common.touch());
-  dispatchHostEvent(element, 'pressOut', EventBuilder.Common.touch());
+export async function longPress(
+  this: UserEventInstance,
+  element: ReactTestInstance,
+  options: PressOptions = { duration: 500 }
+): Promise<void> {
+  await basePress(this.config, element, options);
 }
+
+const basePress = async (
+  config: UserEventInstance['config'],
+  element: ReactTestInstance,
+  options: PressOptions = { duration: 0 }
+): Promise<void> => {
+  // Text and TextInput components are mocked in React Native preset so the mock
+  // doesn't implement the pressability class
+  // Thus we need to call the props directly on the host component
+  const isEnabledHostText =
+    filterNodeByType(element, getHostComponentNames().text) &&
+    isPointerEventEnabled(element) &&
+    !element.props.disabled;
+  const isEnabledTextInput =
+    filterNodeByType(element, getHostComponentNames().textInput) &&
+    isPointerEventEnabled(element) &&
+    element.props.editable !== false;
+
+  if (isEnabledHostText || isEnabledTextInput) {
+    const { onPressIn, onPress, onPressOut } = element.props;
+    await wait(config);
+    if (onPressIn) {
+      onPressIn(EventBuilder.Common.touch());
+    }
+    if (onPress) {
+      onPress(EventBuilder.Common.touch());
+    }
+    await wait(config, options.duration);
+    if (onPressOut) {
+      if (DEFAULT_MIN_PRESS_DURATION - options.duration > 0) {
+        await wait(config, DEFAULT_MIN_PRESS_DURATION - options.duration);
+      }
+      onPressOut(EventBuilder.Common.touch());
+    }
+  }
+
+  if (isEnabledTouchResponder(element)) {
+    await triggerPressEvent(config, element, options);
+    return;
+  }
+
+  const hostParentElement = getHostParent(element);
+  if (!hostParentElement) {
+    return;
+  }
+
+  await basePress(config, hostParentElement, options);
+};
+
+const triggerPressEvent = async (
+  config: UserEventInstance['config'],
+  element: ReactTestInstance,
+  options: PressOptions = { duration: 0 }
+) => {
+  const areFakeTimersEnabled = jestFakeTimersAreEnabled();
+  if (!areFakeTimersEnabled) {
+    warnAboutRealTimers();
+  }
+
+  await wait(config);
+
+  await act(async () => {
+    element.props.onResponderGrant({
+      ...EventBuilder.Common.touch(),
+      dispatchConfig: { registrationName: 'onResponderGrant' },
+    });
+
+    await wait(config, options.duration);
+
+    element.props.onResponderRelease({
+      ...EventBuilder.Common.touch(),
+      dispatchConfig: { registrationName: 'onResponderRelease' },
+    });
+
+    if (DEFAULT_MIN_PRESS_DURATION - options.duration > 0) {
+      await wait(config, DEFAULT_MIN_PRESS_DURATION - options.duration);
+    }
+  });
+};
+
+const isEnabledTouchResponder = (element: ReactTestInstance) => {
+  return (
+    isPointerEventEnabled(element) &&
+    element.props.onStartShouldSetResponder &&
+    element.props.onStartShouldSetResponder()
```

Comment:
optional chaining should work here
```suggestion
    element.props.onStartShouldSetResponder?.()
```